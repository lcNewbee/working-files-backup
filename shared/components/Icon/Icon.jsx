import React from 'react';
import PropTypes from 'prop-types';
import 'font-awesome/css/font-awesome.css';

import utils from '../../utils';
import './Icon.scss';

const propTypes = {
  name: PropTypes.string.isRequired,
  className: PropTypes.string,
  size: PropTypes.oneOf(['lg', '2x', '3x', '4x', '5x']),
  rotate: PropTypes.oneOf(['45', '90', '135', '180', '225', '270', '315']),
  flip: PropTypes.oneOf(['horizontal', 'vertical']),
  fixedWidth: PropTypes.bool,
  spin: PropTypes.bool,
  pulse: PropTypes.bool,
  stack: PropTypes.oneOf(['1x', '2x']),
  inverse: PropTypes.bool,
  Component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
};

const defaultProps = {
  Component: 'span',
};

// 自定义
function isCustomIcon(name) {
  const customNameMap = {
    display: true,
    laptop: true,
    mobile: true,
    mobile2: true,
    tablet: true,
    tv: true,
    meter: true,
    sphere: true,
  };

  return customNameMap[name];
}
function Icon(props) {
  const {
    Component, name, size, rotate, flip, spin,
    stack, inverse, pulse, className, fixedWidth,
  } = props;
  const iconProps = utils.objectAssign({}, props);

  delete iconProps.Component;
  delete iconProps.spin;
  delete iconProps.ref;

  let classNames = 'icon fa';

  // 非自定义图标，Font Awesome图标
  if (!isCustomIcon(name)) {
    if (name) {
      classNames = `${classNames} fa-${name}`;
    }
    if (size) {
      classNames = `${classNames} fa-${size}`;
    }
    if (rotate) {
      classNames = `${classNames} fa-rotate-${rotate}`;
    }
    if (flip) {
      classNames = `${classNames} fa-flip-${flip}`;
    }
    if (fixedWidth) {
      classNames = `${classNames} fa-fw`;
    }
    if (spin) {
      classNames = `${classNames} fa-spin`;
    }
    if (pulse) {
      classNames = `${classNames} fa-pulse`;
    }
    if (stack) {
      classNames = `${classNames} fa-stack-${stack}`;
    }
    if (inverse) {
      classNames = `${classNames} fa-inverse`;
    }

  // 自定义图标
  } else {
    classNames = `${classNames} a-icon-${name}`;
  }

  if (className) {
    classNames = `${classNames} ${className}`;
  }

  return <Component {...iconProps} className={classNames} />;
}

Icon.propTypes = propTypes;
Icon.defaultProps = defaultProps;

export default Icon;
