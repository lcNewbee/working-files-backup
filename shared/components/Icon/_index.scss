@font-face {
  font-family: 'icomoon';
  font-style: normal;
  font-weight: normal;

  src: url('fonts/icomoon.eot?v=0.1.0');
  src: url('fonts/icomoon.eot?#iefix&v=0.1.0') format('embedded-opentype'),
  url('fonts/icomoon.ttf?v=0.1.0') format('truetype'),
  url('fonts/icomoon.woff?v=0.1.0') format('woff'),
  url('fonts/icomoon.svg?v=0.1.0#icomoon') format('svg');
}

.icon {
  vertical-align: middle;
}

[class^="a-icon-"],
[class*=" a-icon-"] {
  /* use !important to prevent issues with browser extensions that change fonts */
  font-family: 'icomoon' !important;
  font-style: normal;
  font-weight: normal;
  font-variant: normal;
  line-height: 1;
  text-transform: none;

  //speak: none;
  /* Better Font Rendering =========== */
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.a-icon-tv:before {
  content: "\e95b";
}
.a-icon-sphere:before {
  content: "\e9c9";
}

.Icon {
  display: inline-block;
  width: 16px;
  height: 16px;
  vertical-align: middle;
  background-repeat: no-repeat;
  background-size: 16px;
}

.Icon-block {
  display: block;
  margin: 0 auto;
}

.Icon-ap-online {
  background-image: url(images/ap-online.png);
}
.Icon-ap-offline {
  background-image: url(images/ap-offline.png);
}
.Icon-wifi-nopass-0 {
  background-image: url(images/icon_wifi_nopass_0.png);
}
.Icon-wifi-nopass-1 {
  background-image: url(images/icon_wifi_nopass_1.png);
}
.Icon-wifi-nopass-2 {
  background-image: url(images/icon_wifi_nopass_2.png);
}
.Icon-wifi-nopass-3 {
  background-image: url(images/icon_wifi_nopass_3.png);
}

.Icon-wifi-pass-0 {
  background-image: url(images/icon_wifi_pass_0.png);
}
.Icon-wifi-pass-1 {
  background-image: url(images/icon_wifi_pass_1.png);
}
.Icon-wifi-pass-2 {
  background-image: url(images/icon_wifi_pass_2.png);
}
.Icon-wifi-pass-3 {
  background-image: url(images/icon_wifi_pass_3.png);
}
