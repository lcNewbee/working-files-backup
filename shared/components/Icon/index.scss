.Icon {
  display: inline-block;
  vertical-align: middle;
  width: 16px;
  height: 16px;
  background-repeat: no-repeat;
  background-size: 16px;
}

.Icon-block {
  display: block;
  margin: 0 auto;
}

.Icon-wifi-nopass-0 {
  background-image: url(images/icons/icon_wifi_nopass_0.png);
}
.Icon-wifi-nopass-1 {
  background-image: url(images/icons/icon_wifi_nopass_1.png);
}
.Icon-wifi-nopass-2 {
  background-image: url(images/icons/icon_wifi_nopass_2.png);
}
.Icon-wifi-nopass-3 {
  background-image: url(images/icons/icon_wifi_nopass_3.png);
}

.Icon-wifi-pass-0 {
  background-image: url(images/icons/icon_wifi_pass_0.png);
}
.Icon-wifi-pass-1 {
  background-image: url(images/icons/icon_wifi_pass_1.png);
}
.Icon-wifi-pass-2 {
  background-image: url(images/icons/icon_wifi_pass_2.png);
}
.Icon-wifi-pass-3 {
  background-image: url(images/icons/icon_wifi_pass_3.png);
}
