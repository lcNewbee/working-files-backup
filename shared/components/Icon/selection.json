{
  "IcoMoonType": "selection",
  "icons": [
    {
      "icon": {
        "paths": [
          "M981.188 288.108c-88.808-12.768-183.382-22.016-282.076-27.22l164.888-164.888-64-64-224.558 224.556c-21.006-0.368-42.156-0.556-63.442-0.556v0l-256-256-64 64 194.196 194.196c-120.922 4.242-236.338 14.524-343.386 29.912-27.532 107.726-42.81 226.752-42.81 351.892s15.278 244.166 42.804 351.89c143.642 20.652 302.34 32.11 469.196 32.11s325.55-11.458 469.188-32.11c27.534-107.724 42.812-226.75 42.812-351.89s-15.278-244.166-42.812-351.892zM863.892 874.594c-107.73 13.766-226.75 21.406-351.892 21.406s-244.166-7.64-351.892-21.406c-20.648-71.816-32.108-151.166-32.108-234.594 0-83.43 11.458-162.78 32.108-234.596 107.726-13.766 226.75-21.404 351.892-21.404 125.136 0 244.162 7.638 351.886 21.404 20.656 71.816 32.114 151.166 32.114 234.596 0 83.428-11.458 162.778-32.108 234.594z"
        ],
        "tags": [
          "tv",
          "television",
          "show"
        ],
        "defaultCode": 59739,
        "grid": 16,
        "attrs": []
      },
      "attrs": [],
      "properties": {
        "ligatures": "tv, television",
        "name": "tv",
        "order": 3,
        "id": 92,
        "prevSize": 32,
        "code": 59739
      },
      "setIdx": 1,
      "setId": 1,
      "iconIdx": 91
    },
    {
      "icon": {
        "paths": [
          "M512 1024c282.769 0 512-229.231 512-512 0-282.771-229.231-512-512-512-282.771 0-512 229.229-512 512 0 282.769 229.229 512 512 512zM222.304 341.333c-9.152 43.151-14.929 88.951-16.809 136.533h-135.942c3.558-47.142 14.532-92.922 32.725-136.533h120.026zM801.696 682.667c9.152-43.151 14.929-88.951 16.809-136.533h135.94c-3.558 47.14-14.53 92.922-32.723 136.533h-120.026zM292.229 682.667c-10.255-43.693-16.425-89.444-18.432-136.533h204.070v136.533h-185.638zM477.867 750.933v199.827c-15.563-4.529-30.98-12.128-46.11-22.795-28.354-19.989-55.479-50.782-78.443-89.052-15.874-26.455-29.602-55.91-41.062-87.983h165.615zM670.686 838.916c-22.963 38.27-50.086 69.065-78.443 89.052-15.127 10.667-30.547 18.266-46.11 22.795v-199.829h165.615c-11.462 32.070-25.19 61.525-41.062 87.983zM546.133 682.667v-136.533h204.068c-2.005 47.089-8.177 92.841-18.43 136.533h-185.638zM921.719 341.333c18.193 43.612 29.165 89.391 32.723 136.533h-135.94c-1.879-47.582-7.657-93.382-16.809-136.533h120.026zM750.202 477.867h-204.068v-136.533h185.638c10.253 43.691 16.425 89.444 18.43 136.533zM546.133 273.067v-199.829c15.561 4.529 30.98 12.13 46.11 22.795 28.356 19.989 55.482 50.784 78.443 89.052 15.874 26.458 29.6 55.913 41.065 87.983h-165.617zM353.314 185.084c22.963-38.268 50.089-69.062 78.443-89.052 15.13-10.665 30.547-18.266 46.11-22.795v199.829h-165.617c11.462-32.070 25.19-61.525 41.065-87.983zM477.867 341.333v136.533h-204.068c2.005-47.089 8.177-92.843 18.432-136.533h185.636zM205.495 546.133c1.879 47.582 7.657 93.382 16.809 136.533h-120.026c-18.193-43.612-29.167-89.393-32.725-136.533h135.942zM137.965 750.933h102.293c19.881 62.679 47.334 117.647 80.337 161.523-45.372-21.696-86.462-50.79-122.362-86.69-22.912-22.91-43.053-47.934-60.269-74.833zM825.766 825.766c-35.902 35.902-76.992 64.994-122.362 86.69 33.005-43.876 60.454-98.844 80.337-161.523h102.291c-17.214 26.899-37.357 51.923-60.267 74.833zM886.033 273.067h-102.291c-19.883-62.679-47.334-117.647-80.339-161.525 45.372 21.696 86.462 50.79 122.362 86.692 22.912 22.91 43.055 47.934 60.269 74.833zM198.234 198.234c35.9-35.902 76.992-64.996 122.362-86.692-33.005 43.876-60.457 98.846-80.337 161.525h-102.293c17.216-26.899 37.357-51.923 60.269-74.833z"
        ],
        "tags": [
          "sphere",
          "globe",
          "internet"
        ],
        "defaultCode": 59849,
        "grid": 16,
        "attrs": []
      },
      "attrs": [],
      "properties": {
        "ligatures": "sphere, globe",
        "name": "sphere",
        "order": 12,
        "id": 202,
        "prevSize": 32,
        "code": 59849
      },
      "setIdx": 1,
      "setId": 1,
      "iconIdx": 199
    }
  ],
  "height": 1024,
  "metadata": {
    "name": "icomoon"
  },
  "preferences": {
    "showGlyphs": true,
    "showQuickUse": true,
    "showQuickUse2": true,
    "showSVGs": true,
    "fontPref": {
      "prefix": "a-icon-",
      "metadata": {
        "fontFamily": "icomoon",
        "majorVersion": 1,
        "minorVersion": 0
      },
      "metrics": {
        "emSize": 1024,
        "baseline": 6.25,
        "whitespace": 50
      },
      "embed": false,
      "showSelector": false,
      "showMetrics": false,
      "showMetadata": false,
      "showVersion": true
    },
    "imagePref": {
      "prefix": "icon-",
      "png": true,
      "useClassSelector": true,
      "color": 0,
      "bgColor": 16777215,
      "classSelector": ".icon"
    },
    "historySize": 100,
    "showCodes": true,
    "gridSize": 16,
    "showLiga": true,
    "showGrid": true
  }
}