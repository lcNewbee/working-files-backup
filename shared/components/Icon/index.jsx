import React, { PropTypes } from 'react';
import 'font-awesome/css/font-awesome.css';
import utils from '../../utils';
import './_index.scss';

const propTypes = {
  name: PropTypes.string.isRequired,
  className: PropTypes.string,
  size: PropTypes.oneOf(['lg', '2x', '3x', '4x', '5x']),
  rotate: PropTypes.oneOf(['45', '90', '135', '180', '225', '270', '315']),
  flip: PropTypes.oneOf(['horizontal', 'vertical']),
  fixedWidth: PropTypes.bool,
  spin: PropTypes.bool,
  pulse: PropTypes.bool,
  stack: PropTypes.oneOf(['1x', '2x']),
  inverse: PropTypes.bool,
  Component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
};

const defaultProps = {
  Component: 'span',
};

// 自定义
function isCustomIcon(name) {
  const customNameMap = {
    display: true,
    laptop: true,
    mobile: true,
    mobile2: true,
    tablet: true,
    tv: true,
    meter: true,
    sphere: true,
  };

  return customNameMap[name];
}

function Icon(props) {
  const {
    Component,
    name, size, rotate, flip, spin, fixedWidth, stack, inverse,
    pulse, className,
  } = props;
  const iconProps = utils.objectAssign({}, props);
  delete iconProps.Component;
  delete iconProps.spin;

  let classNames = 'icon fa';

  // 非自定义图标，Font Awesome图标
  if (!isCustomIcon(name)) {
    if (name) {
      classNames = `${classNames} fa-${name}`;
    }
    if (size) {
      classNames = `${classNames} fa-${size}`;
    }
    if (rotate) {
      classNames = `${classNames} fa-rotate-${rotate}`;
    }
    if (flip) {
      classNames = `${classNames} fa-flip-${flip}`;
    }
    if (fixedWidth) {
      classNames = `${classNames} fa-fw`;
    }
    if (spin) {
      classNames = `${classNames} fa-spin`;
    }
    if (pulse) {
      classNames = `${classNames} fa-pulse`;
    }
    if (stack) {
      classNames = `${classNames} fa-stack-${stack}`;
    }
    if (inverse) {
      classNames = `${classNames} fa-inverse`;
    }

  // 自定义图标
  } else {
    classNames = `${classNames} a-icon-${name}`;
  }

  if (className) {
    classNames = `${classNames} ${className}`;
  }

  return <Component {...iconProps} className={classNames} />;
}

Icon.propTypes = propTypes;
Icon.defaultProps = defaultProps;

export default Icon;
