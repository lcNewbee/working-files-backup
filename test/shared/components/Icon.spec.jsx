// Link.react-test.js
import React from 'react';
import Icon from 'shared/components/Icon';
import renderer from 'react-test-renderer';

test('Link changes the class when hovered', () => {
  const component = renderer.create(
    <Icon name="cog" spin size="2x" />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // // manually trigger the callback
  // tree.props.onMouseEnter();
  // // re-rendering
  // tree = component.toJSON();
  // expect(tree).toMatchSnapshot();

  // // manually trigger the callback
  // tree.props.onMouseLeave();
  // // re-rendering
  // tree = component.toJSON();
  // expect(tree).toMatchSnapshot();
});
