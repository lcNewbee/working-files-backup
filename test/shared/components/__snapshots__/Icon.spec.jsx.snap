exports[`test Link changes the class when hovered 1`] = `
<span
  className="icon fa fa-cog fa-2x fa-spin"
  name="cog"
  size="2x" />
`;
